// @medicomp/medcin-client/medcinResponseItem.js
/*** Created by pkarcher on 1/19/2015.*/
'use strict';

// var medicomp = medicomp || {};

var ResponseItem = function () {

  this.medcinId = null;
  this.prefix = null; 
  this.modifier = null;
  this.status = null;
  this.result = null;
  this.nodeKey = null;
  this.description = null;
  this.termType = null;
  this.flags = null;
  this.gpFlags = null;
  this.itemFlags = null;
  this.subs = null;
  this.varData = null;
  this.isResponseItem = function() {
    return true;
  };
  this.toString = function() {
    return 'ResponseItem ' + this.medcinId + ': ' + this.description;
  };
  
  //return {
  //  medcinId : null,
  //  prefix : null,
  //  modifier : null,
  //  status : null,
  //  result : null,
  //  nodeKey : null,
  //  description : null,
  //  termType : null,
  //  flags : null,
  //  gpFlags : null,
  //  itemFlags : null,
  //  subs : null,
  //  varData : null,
  //  toString : function() {
  //    return 'ResponseItem ' + this.medcinId + ': ' + this.description;
  //  }
  //};

};

module.exports = ResponseItem;
