// @medicomp/medcin-client/ServerCodes.js
'use strict';


//class ServerCode #@private :nodoc:

function serverCodes () {

  return {
    MedcinInit : 1,
    MedcinEnd : 2,
    SetDebugOutput : 3,
    SetDataDir : 4, //Not in integration tests.
    RegisterCallBack : 5, //Not implemented; not found in .NET client
    ClearChart : 6, //deprecated in documentation, but implemented because it was easy.
    AddChart : 7,
    RemoveChart : 8,
    Calculate : 9,
    HaveList : 10,
    Activate : 11,
    GetIPDx : 12,
    AddDx : 13,
    ClearDx : 14,
    ActivateDx : 15,
    HaveDx : 16,
    ActivateExpansion : 17,
    WordSearch : 18,
    HaveWordSearch : 19,
    ActivateWordSearch : 20,
    RangeNormalHigh : 21,
    RangeNormalLow : 22,
    UnitGuess : 23,
    TestValueConvert : 24,
    UnitEnum : 25,
    CalcScale : 26,
    QualifierList : 27,
    QualifierDesc : 28,
    NodeKey : 29,
    TermType : 30,
    Specialty : 31,
    Unit : 32,
    Search : 33,
    SVoice : 34,
    Nomin : 35,
    NVoice : 36,
    SDocPos : 37,
    SDocNeg : 38,
    SPatPos : 39,
    SPatNeg : 40,
    PDocPos : 41,
    PDocNeg : 42,
    PPatPos : 43,
    PPatNeg : 44,
    RxCode : 45,
    TestCode : 46,
    RiskCode : 47,
    Privacy : 48,
    Complex : 49,
    UDFCodeList : 50,
    ExpandCodeHint : 51,
    GetProperty : 52,
    NodeLevel : 56,
    ParaFlow : 57,
    SentFlow : 58,
    Title : 59,
    Subs : 60,
    ParentMedcinId : 61,
    Flags : 62,
    EFlags : 63,
    Management : 64,
    AFlags : 65,
    CAFlags : 66,
    QFlags : 67,
    StartAge : 68,
    Age : 69,
    AddUserFinding : 70,
    UserFindingCodes : 71,
    UserFindingRange : 72,
    Transcribe : 73,
    TranscribeFlag : 74,
    Narrate : 75,
    TranscribeItem : 76,
    AddNarrate : 77,
    AddEpisode : 78,
    AddValue : 79,
    AddNote : 80,
    AddRxStr : 81,
    ClearNarrative : 82,
    DeleteNarrate : 83,
    AddDocName : 84,
    AddDemographics : 85,
    AddEncounter : 86,
    GetGroupSequence : 87,
    SetGroupSequence : 88,
    GetGroupNumber : 89,
    GroupName : 90,
    GetNarrativeFormat : 91,
    SetNarrativeFormat : 92,
    GetNarrativeOutput : 93,
    SetNarrativeOutput : 94,
    GetNarrativeContext : 95,
    SetNarrativeContext : 96,
    GetTimeFormat : 97,
    SetTimeFormat : 98,
    GetTimedSections : 99,
    SetTimedSections : 100,
    RTFStyle : 101,
    ParagraphStyle : 102,
    AddItemTag : 103,
    SetShowItemTag : 104,
    GetShowItemTag : 105,
    ItemSTag : 106,
    ItemLTag : 107,
    SetShowItemId : 108,
    SetShowMedcinId : 109,
    GetVersion : 110,
    ReturnData : 113,
    MedcinCallBack : 114,
    SetGroupNumber : 115,
    ClientCache : 116,
    AddDxPrompt : 117,
    SetPatientGeneric : 118,
    SetPatient : 119,
    ClearChartFindings : 120,
    GetMessageString : 121,
    SetMessageString : 122,
    GetNarrateOptions : 123,
    SetNarrateOptions : 124,
    NarrateEx : 125,
    GetStateKey : 126,
    RestoreState : 127,
    NarrateGroups : 128,
    ActivateEm : 129,
    ActivateEmPrompt : 130,
    CalculateEm : 131,
    ActivateExpansionEx : 132,
    LinkFindings : 133,
    InfoLathd : 134,
    BuildListMode : 138,
    GetUserFindingTag : 142,
    SetUserFindingTag : 143,
    TSearch : 144,
    STersePos : 145,
    STerseNeg : 146,
    PTersePos : 147,
    PTerseNeg : 148,
    DoseList : 149,
    BrandList : 150,
    Dose : 151,
    HistFlush : 152,
    HistNarrate : 153,
    HistTranscribeItem : 154,
    HistAddNarrate : 155,
    HistAddEpisode : 156,
    HistAddValue : 157,
    HistAddNote : 158,
    HistAddRxStr : 159,
    HistDelete : 160,
    HistLinkFindings : 162,
    HistAddDocName : 163,
    HistAddEncounter : 164,
    HistPacket : 165,
    HistActivate : 166,
    HistActivateDx : 167,
    HistLocate : 168,
    HistActivateWordSearch : 169,
    HistAddChart : 170,
    HistAddDx : 171,
    HistCalculate : 172,
    HistClearChart : 173,
    HistClearChartFindings : 174,
    HistClearDx : 175,
    HistHaveDx : 176,
    HistHaveList : 177,
    HistHaveWordSearch : 178,
    HistReleasePool : 179,
    HistRemoveChart : 180,
    HistRetrieveEncounter : 181,
    HistRetrieveOrder : 182,
    HistWordSearch : 183,
    HistSetPoolId : 184,
    HistSetPoolPid : 185,
    HistCreatePool : 186,
    HistAddOrder : 187,
    HistDeleteEncounter : 188,
    HistEncounterList : 189,
    HistAddFuList : 190,
    GetBuildListMode : 191,
    GetGroupSequenceEx : 192,
    SetGroupSequenceEx : 193,
    SetDoctorID : 194,
    GetDoctorName : 195,
    SaveUdf : 196,
    NextUdf : 197,
    RtfIndentLevel : 198,
    HistSetDoctorId : 199,
    ImageSupport : 200,
    GetImageId : 201,
    GetImageLen : 202,
    GetImage : 203,
    ImgGetHotspotsLen : 204,
    GetHotSpots : 205,
    GetListItemsLen : 206,
    GetListItems : 207,
    AddDecorationStyle : 208,
    SetDecorationStyle : 209,
    AddNarrate2 : 210,
    AddNarrate2Ex : 211,
    HistGroupNumber : 212,
    HistSetDecorationStyle : 213,
    HistAddNarrate2 : 214,
    HistAddNarrate2Ex : 215,
    MedispanIds : 216,
    HistList : 217,
    SepID : 218,
    MedispanEx : 219,
    CheckAlert : 220,
    EM95Cat : 221,
    GpFlags : 222,
    ICDMan : 223,
    EM95Item : 224,
    GetSubGroupSequence : 225,
    SetSubGroupSequence : 226,
    GetImageIdEx : 227,
    GetReferenceInfo : 228,
    CodeFromMedcin : 229,
    SetCodeDate : 230,
    SetAlertLookBack : 231,
    SetAlertSuppress : 232,
    FindingCgn : 233,
    FindingTgn : 234,
    BaseNarGroup : 235,
    GetItemData : 236,
    HistGetItemData : 237,
    ScreenForCodes : 238,
    AddSequence : 239,
    AddSequenceEx : 240,
    GetSequence : 241,
    HistAddSequence : 242,
    SetRtfStyles : 243,
    SetParagraphStyles : 244,
    SetChapter : 245,
    AddChapter : 246,
    HistAddChapter : 247,
    ImageSearch : 248,
    CodeFromItem : 249,
    HistCodeFromItem : 250,
    GetImageProperties : 251,
    GetSubGroup : 252,
    CodeFromParms : 253,
    HistAddRangeScale : 254,
    HistGetGroupNumber : 255,
    FormatDate : 256,
    HistGridLayout : 257,
    HistDataFilter : 258,
    AddSpecifier : 259,
    HistAddSpecifier : 260,
    BaseUserFinding : 261,
    AddEMData : 262,
    AddFindingImage : 263,
    HistAddFindingImage : 264,
    UserFindingProperties : 265,
    GetCaption : 266,
    MedcinIdFromNodekey : 267,
    AddNarrateCode : 268,
    HistAddNarrateCode : 269,
    GetCodeDescription : 270,
    AddEncounterHeadings : 271,
    HistAddEncounterHeadings : 272,
    RegisterClient : 273,
    GetServerProperties : 274,
    SetSubGroup : 275,
    HistSetSubGroup : 276,
    Cost : 277
  };

}

module.exports = serverCodes();
